iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
+    removeIndexFromMap(m_instanceMap, index);
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const capro::IdString_t& service,
-                           const capro::IdString_t& instance) const
+                           const capro::IdString_t& instance) const noexcept
 {
-    if (instance == Wildcard)
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != Wildcard && service != Wildcard)
+    {
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
+                                possibleServices.end(),
+                                possibleInstances.begin(),
+                                possibleInstances.end(),
+                                ::std::back_inserter(intersection));
+
+        for (auto& value : intersection)
+        {
+            searchResult.push_back(m_serviceDescriptionVector[value]);
+        }
+    }
+    // Find (*, K2)
+    // O(log n + #result)
+    else if (service == Wildcard && instance != Wildcard)
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        auto range = m_instanceMap.equal_range(instance);
+        for (auto entry = range.first; entry != range.second; ++entry)
         {
-            instances.push_back(instance);
+            searchResult.push_back(m_serviceDescriptionVector[entry->second]);
         }
     }
-    else
+    // Find (K1, *)
+    // O(log n + #result)
+    else if (instance == Wildcard && service != Wildcard)
     {
-        auto& instanceSet = m_serviceMap[service].instanceSet;
-        auto iter = std::find(instanceSet.begin(), instanceSet.end(), instance);
-        if (iter != instanceSet.end())
+        auto range = m_serviceMap.equal_range(service);
+        for (auto entry = range.first; entry != range.second; ++entry)
         {
-            instances.push_back(*iter);
+            searchResult.push_back(m_serviceDescriptionVector[entry->second]);
         }
     }
+    else
+    {
+        // Find (*, *)
+        // O(1)
+        searchResult = m_serviceDescriptionVector;
```

Comment:
In general the question arises whether we do always want to pass the result by copy, especially in the all wildcard case. Those will be many matches and they are not fast to copy if the result is of the order 1000.

Reply:
We'll see how things develop and measure once we have the first implementations ready. Then we can iterate over it again. My personal mantra: Make it work, make it fast, make it beautiful.